src/utils/src/ThreadsafeBlockingQueue.c

```diff
@@ -0,0 +1,214 @@
+#include "Include_i.h"
+
+/**
+ * Create a threadsafe blocking queue
+ */
+STATUS safeBlockingQueueCreate(PSafeBlockingQueue* ppSafeQueue)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PSafeBlockingQueue pSafeQueue = NULL;
+
+    CHK(ppSafeQueue != NULL, STATUS_NULL_ARG);
+
+    // Allocate the main structure
+    pSafeQueue = (PSafeBlockingQueue) MEMCALLOC(1, SIZEOF(SafeBlockingQueue));
+    CHK(pSafeQueue != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    ATOMIC_STORE_BOOL(&pSafeQueue->terminate, FALSE);
+
+    pSafeQueue->mutex = MUTEX_CREATE(FALSE);
+    CHK_STATUS(semaphoreEmptyCreate(INT32_MAX, &(pSafeQueue->semaphore)));
+    CHK_STATUS(stackQueueCreate(&(pSafeQueue->queue)));
+
+    *ppSafeQueue = pSafeQueue;
+
+CleanUp:
+    if (STATUS_FAILED(retStatus) && pSafeQueue != NULL) {
+        SAFE_MEMFREE(pSafeQueue);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Frees and de-allocates the thread safe blocking queue
+ */
+STATUS safeBlockingQueueFree(PSafeBlockingQueue pSafeQueue)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+
+    // set terminate flag, lock mutex -- this assures all other threads
+    // are no longer directly interacting with the queue
+    //
+    // free semaphore, this will unblock any threads blocking on said semaphore,
+    // and then exit from the terminate flag
+    //
+    // unlock mutex
+    ATOMIC_STORE_BOOL(&pSafeQueue->terminate, TRUE);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    CHK_STATUS(semaphoreFree(&(pSafeQueue->semaphore)));
```

Comment:
This will leak if the semaphore fails to free. Perhaps remembering the status code of the operation and/or DLOGWing it?

Reply:
I meant will leak the mutex and the queue

Reply:
Missed the lock problem since semaphoreFree() can only fail if the semaphore_handle is NULL, which would mean the constructor failed.

With that in mind I'm removing the CHK_STATUS on both of the Free() calls here. The destructors of the utility objects only fail (and should only fail) in the event of a null object, so the check doesn't offer anything.

Reply:
This is more about resilience of the code. In the future the failure modes can be more diverse so accounting for it is a good idea.